### ISSUE TITLE
Zerotier Network Throughput

### ISSUE THREAD

**Issue by blurredpixel** [zerotier/ZeroTierOne#445]
Going from a host with 100/100 fiber to a EC2 instance, this is the throughput I am getting. Is this expected or is it a part of the upgrade process? 

![image](https://cloud.githubusercontent.com/assets/12002376/22816642/35a0fdbe-ef31-11e6-9e2d-fe5023c635ac.png)

**Comment by blurredpixel** [comment]
Here is a speedtest from ECC2 to Charlotte, NC

![image](https://cloud.githubusercontent.com/assets/12002376/22816750/c2b0eea8-ef31-11e6-83ca-c040a5ae865b.png)

**Comment by adamierymenko** [comment]
Eww. Time for some profiling.

Do you get better with 1.1.14?

I do have a confession to make though: we are following a "premature optimization is the root of all evil" philosophy with ZeroTier and focusing on getting its security, functionality, ease of use, and reliability nailed before truly delving into speed. As it stands we actually haven't spent a ton of time optimizing it, and there are many opportunities in the code to eliminate e.g. unnecessary memcpy. We will get there.

But yes we do want at least acceptable performance and this is icky.

**Comment by blurredpixel** [comment]
@adamierymenko 
I never noticed throughput 'issues' until I've started using ZT as the backbone for my test labs connecting together, and I moved all of my main VMs to a host at work we use as a sandbox. Once it was on a dedicated connection, I started seeing poor SMB performance across the ZTWAN. 

Here's some good news! 
Desktop (running ZT client that's also on same physical LAN as a ZT subnet) > virtual router > server
I basically got 20.5/MBs through zerotier when it was on the same physical LAN. Seems to only be when traversing the internet that the speed hit takes place. I really wish I knew enough C to dive into the code to see how all the encapsulation works, as I would dedicate any free time I have to this project because it's so amazing.

**Comment by adamierymenko** [comment]
Hmm... what does `listpeers` say? Do you have direct paths? Maybe you're relaying.

**Comment by blurredpixel** [comment]
![image](https://cloud.githubusercontent.com/assets/12002376/22848859/73acced2-efc5-11e6-953d-994febeee24e.png)

**Comment by adamierymenko** [comment]
Which one are you talking to? 6423... has no direct path, the rest do. (One, e5cd, is one of our network controllers and doesn't count.)

**Comment by blurredpixel** [comment]
I think i'm starting to see what you're getting at.... what could cause it? 

![image](https://cloud.githubusercontent.com/assets/12002376/22848974/a25d47ce-efc6-11e6-86a8-fc0c64bad073.png)

**Comment by adamierymenko** [comment]
Possibly an EC2 security policy issue. Try opening UDP/9993 explicitly.

**Comment by blurredpixel** [comment]
The ones that are showing up as "0" aren't on EC2. Hmmmm let me mess around with some FW settings and I'll report back and see if that status changes

**Comment by blurredpixel** [comment]
Okay so I tweaked some FW settings and I don't see a difference here. I'm not understanding why some of the peers will show up, then from some hosts they show up with the -0-

![image](https://cloud.githubusercontent.com/assets/12002376/22849129/1ee12706-efc8-11e6-8b2c-417bad9b2ec8.png)

**Comment by adamierymenko** [comment]
Is the one with no direct path one of the slow ones?

**Comment by blurredpixel** [comment]
So I was running the tests from (in the picture) fncloudedge to aptedge 

fncloudedge isn't even showing aptedge in its peers list; however, aptedge is peering with with it.

**Comment by blurredpixel** [comment]
Are the root ZT servers running 1.0.5? Could that be it?

**Comment by adamierymenko** [comment]
They are, but all they do is introduce and relay packets. Wouldn't be related. We haven't updated them because no need.

**Comment by blurredpixel** [comment]
None of the zt clients are showing as related in my.zerotier btw

**Comment by adamierymenko** [comment]
Have to run but if the ones that are slow lack direct paths, this means the slowness is because you are relaying. I'd check firewall settings everywhere: on local hosts (iptables), AWS, etc.

**Comment by blurredpixel** [comment]
Alright I'll give everything a once over. Thanks for your input!

**Comment by blurredpixel** [comment]
EC2 to my APT (which isn't in peers list) saturates EC2 and my APT links. Something weird with that one client, and since it hosts file shares, AD, etc. i really notice the speed difference. Will keep investigating.

**Comment by orware** [comment]
I have a related question if you don't mind me adding to this older thread :-).

I have two Windows machines running Zerotier (one is a Windows Server 2019 VM and the other is a Windows 10 Workstation).

Both machines are inside our internal network so the max bandwidth available between the two machines would be 1 Gbit and there aren't any firewalls enabled on either machine (the Windows Firewall is disabled to rule that out as an issue).

Running iperf using the normal IP address of the server machine over our internal network results in a speed of 904 Mbits/sec.

Running iperf using the Zerotier IP address of the server machine results in a speed of about 214 Mbit/sec.

One possible explanation would be slowdowns related to the encryption process and security features, but I seemed to recall that the impact of those should be relatively light so I don't think a slowdown that dramatic is expected so I wanted to reach out to see if there's anything else I might be able to try out on my end and report back on.

**Comment by blurredpixel** [comment]
@orware That's about the speed I get 2 years later after I created this  issue. Whether it's over the public internet or even LAN, it tops out around 25MB/s

**Comment by orware** [comment]
@blurredpixel,

I came across this Zerotier blog post (which I think I had read before at some point) yesterday which had some statistics as well:
https://www.zerotier.com/blog/2017-04-20-benchmarks.shtml

From that blog post's controlled settings, it looks like there's a 90% drop in bandwidth from the maximum speed shown (4760 mbps) compared to the observed Zerotier performance (484 mbps) so maybe the performance we're seeing is what's to be expected? I say that since the observed speed I mentioned above is not too far from that reduction (and it's actually pretty accurate for a different test I ran between the on campus server machine and an off campus cloud VM that likely has a 1 gbps max speed itself since in that case the bandwidth between the two machines ended up being about 126 mbps). In that case it would match up pretty closely with Azure's Basic VPN Tier speeds of 100 mbps (https://azure.microsoft.com/en-us/pricing/details/vpn-gateway/).
